include/dlaf/communication/sync/broadcast.h

```diff
@@ -55,10 +57,29 @@ void receive_from(const int broadcaster_rank, Communicator& communicator, DataOu
 }
 }
 
+/// Helper function for preparing a tile for sending.
+///
+/// Duplicates the tile to CPU memory if CUDA RDMA is not enabled for MPI.
+/// Returns the tile unmodified otherwise.
+template <Device D, typename T>
+auto prepareSendTile(hpx::shared_future<matrix::Tile<T, D>> tile) {
+  return matrix::duplicateIfNeeded<CommunicationDevice<D>::value>(std::move(tile));
+}
+
+/// Helper function for handling a tile after sending.
+///
+/// If CUDA RDMA is disabled, the tile returned from recvTile will always be on
+/// the CPU. This helper duplicates to the GPU if the first template parameter
+/// is a GPU device. The first template parameter must be given.
+template <Device D, typename T>
+auto handleRecvTile(hpx::future<matrix::Tile<T, CommunicationDevice<D>::value>> tile) {
+  return matrix::duplicateIfNeeded<D>(std::move(tile));
+}
+
```

Comment:
I agree! There's no particular reason for them to not be in `internal`. The main reason they're explicitly called inside the algorithm, rather than somewhere hidden is that the "wrapper" send/recv functions that used to call dataflow were lifted up so that dataflow is called explicitly in the algorithm. The only restriction is that `handleRecvTile/prepareSendTile` should ideally be outside the dataflow call.

`handleRecvTile` could in principle be inside `recvTile` as a `return dataflow(copyexecutor, handleRecvTile, ...)` at the end, but the future from `prepareSendTile` should really be passed to `dataflow(mpiexecutor, sendTile, ...)` instead of being called inside `sendTile`

I'm open to what you all think is clearest, but it sounds to me right now like I should reintroduce the wrapper functions that call dataflow.

Reply:
After your offline explanation of the restrictions I see your point.

IMHO in general we should try to prefer "schedulable" functions instead of "scheduling" ones (for reference see https://github.com/eth-cscs/DLA-Future/pull/318#issuecomment-767350611).

I think that with the new needs  of `handle/prepare` for communications, plus the fact that we always want to use the same executor for communication (right?), it is reasonable to make them "scheduling" functions, even if we will have to investigate if there is an easy way to customize annotation for them (e.g. "send diagonal tile row-wise" instead of a generic "send tile").

@msimberg @rasolca @teonnik from an API point of view, should we make it clear that they inside call `dataflow`, so they shall not go into a task?

**TL;DR** I agree you need to reintroduce the wrapper functions that calls dataflow (what I refer to as "scheduling" functions).

Reply:
Well, I renamed the wrappers to `scheduleSendTile`/`scheduleRecv(Alloc)Tile` now. Alternatively I'd simply call them `sendTileWrapper` etc. I'm sure we'll come up with a better naming scheme in the future. Some potential alternatives:

- Move the "task" functions to the `internal` namespace to indicate that they shouldn't be called directly and call the wrappers plain `recv/sendTile`. I didn't move them now since they are still used in non-GPUized algorithms. The fact that the wrappers take an executor as the first argument is a good indication that they shouldn't be used directly as tasks.
- Somehow indicate the RDMA-ness of the communication functions. This is in principle orthogonal to whether or not they should be tasks in themselves, but may help clarify how they should be used. E.g. something like `recvTileRDMA` for the plain version that always receives on the target device and `recvTileViaCPU` for  a version that receives via the CPU (conditionally or not?).

Unless someone gets a big a-ha moment from these ideas I suggest we go ahead with something relatively generic like the `schedule*` or `*Wrapper` names.